File: framework/php/src/Soda/DS/SortedMap.php
<?php
namespace Soda\DS;

require_once __DIR__ . '/SkipList.php';

class SortedMap
{
    private SkipList $skipList;

    function __construct($cmp = null)
    {
        if ($cmp == null) {
            $cmp = fn($a, $b) => $a - $b;
        }
        $this->skipList = new SkipList(10, 0.5, $cmp);
    }

    function has($key): bool
    {
        return $this->skipList->query($key) != null;
    }

    function get($key, $default = null)
    {
        $node = $this->skipList->query($key);
        return $node != null ? $node->value : $default;
    }

    function size(): int
    {
        return $this->skipList->size();
    }

    function set($key, $value)
    {
        $this->skipList->update($key, $value);
    }

    function del($key)
    {
        $this->skipList->remove($key);
    }

    function lowerBound($key): ?array
    {
        return $this->skipList->lowerBound($key);
    }

    function iterate(): \Generator
    {
        yield from $this->skipList->items();
    }

}

File: framework/go/src/ds/sortedmap.go
package ds

type SortedMap struct {
    skipList *SkipList
}

func NewSortedMap(cmpFunc interface{}) *SortedMap {
    return &SortedMap{
        skipList: NewSkipList(10, 0.5, cmpFunc),
    }
}

func (sm *SortedMap) Has(key interface{}) bool {
    return sm.skipList.Query(key) != nil
}

func (sm *SortedMap) GetOrDefault(key interface{}, defaultVal interface{}) interface{} {
    node := sm.skipList.Query(key)
    if node != nil {
        return node.Value
    } else {
        return defaultVal
    }
}

func (sm *SortedMap) Get(key interface{}) interface{} {
    return sm.GetOrDefault(key, nil)
}

func (sm *SortedMap) Size() int {
    return sm.skipList.Size()
}

func (sm *SortedMap) Set(key interface{}, value interface{}) {
    sm.skipList.Update(key, value)
}

func (sm *SortedMap) Del(key interface{}) {
    sm.skipList.Remove(key)
}

func (sm *SortedMap) LowerBoundKey(key interface{}) interface{} {
    node := sm.skipList.LowerBound(key)
    if node != nil {
        return node.Key
    }
    return nil
}

func (sm *SortedMap) Items() [][]interface{} {
    var items [][]interface{}
    for _, node := range sm.skipList.Items() {
        items = append(items, []interface{}{node.Key, node.Value})
    }
    return items
}
